documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1182 @@
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Reverses the sign of each component of the quaternion.</summary>
+    /// <param name="value">The quaternion to negate.</param>
+    /// <returns>The negated quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static Quaternion<T> Negate<T>(Quaternion<T> value)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Divides each component of a specified <see cref="Quaternion{T}" /> by its length.</summary>
+    /// <param name="value">The quaternion to normalize.</param>
+    /// <returns>The normalized quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static Quaternion<T> Normalize<T>(Quaternion<T> value)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Interpolates between two quaternions, using spherical linear interpolation.</summary>
+    /// <param name="quaternion1">The first quaternion.</param>
+    /// <param name="quaternion2">The second quaternion.</param>
+    /// <param name="amount">The relative weight of the second quaternion in the interpolation.</param>
+    /// <returns>The interpolated quaternion.</returns>
+    public static Quaternion<T> Slerp<T>(Quaternion<T> quaternion1, Quaternion<T> quaternion2, T amount)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>, IComparisonOperators<T, T, bool>;
+
+    /// <summary>Subtracts each element in a second quaternion from its corresponding element in a first quaternion.</summary>
+    /// <param name="value1">The first quaternion.</param>
+    /// <param name="value2">The second quaternion.</param>
+    /// <returns>The quaternion containing the values that result from subtracting each element in <paramref name="value2" /> from its corresponding element in <paramref name="value1" />.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static Quaternion<T> Subtract<T>(Quaternion<T> value1, Quaternion<T> value2)
+        where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+    
+    /// <summary>Calculates the length of the quaternion.</summary>
+    /// <returns>The computed length of the quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static T Length<T>(this Quaternion<T> self) where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+
+    /// <summary>Calculates the squared length of the quaternion.</summary>
+    /// <returns>The length squared of the quaternion.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static T LengthSquared<T>(this Quaternion<T> self) where T : IRootFunctions<T>, ITrigonometricFunctions<T>;
+}
+```
+
+# Geometric Types
+
+The following Geometric Types are defined:
+- Box2DF
+- Box2DI
+- Box3DF
+- Box3DI
+- CircleF
+- CircleI
+- CubeF
+- CubeI
```

Comment:
Really this is a rectangular prism, not a cube.

Reply:
The 2.X type was called Cube, are we changing that?

Reply:
We're changing everything, so now is a great time to decide on better names if there are any.

Reply:
I feel like this may be a bit too much bikeshedding and also removing some familiarity without great reasoning for it. Our Rectangle could also be a Square, just like Cube could be a Cuboid. I don’t have strong feelings here, and because of this I think familiarity (for me) is overriding.

Reply:
Aside from just Silk familiarity, Cube is pretty commonly used for any AABB. Unreal uses this terminology, where as Unity uses Box which we already have used for the D3D representation.
On a side note, I don't think the logic of a rectangle can be a square is a good argument as mathematically a square is a rectangle, a cuboid is not a cube by definition, but the reverse is true.

Reply:
If you're tracking things like `origin` (the position) in addition to the `size`, then `AABB` simply drops the `rotation` field of the transform

GLM and DXM define `BoundingBox` as its own thing, since its often used in physics: https://learn.microsoft.com/en-us/windows/win32/api/DirectXCollision/ns-directxcollision-boundingbox

They also have variants (like `BoundingOrientedBox`, which is AABB) that drop fields for this efficiency purpose.